### PR TITLE
Redshfit: sync tables with partial select permission

### PR DESCRIPTION
**Pull request by qnkhuat**
Similiar to https://github.com/metabase/metabase/pull/40034 but for redshift

### CODE REVIEW

File: modules/drivers/redshift/src/metabase/driver/redshift.clj [hunk 1]
```diff
@@ -433,30 +433,33 @@
...
+        (str/join
+         "\n"
+         ["with table_privileges as ("
+          " select"
+          "   NULL as role,"
+          "   t.schemaname as schema,"
+          "   t.objectname as table,"
+          ;; interestingly if you have table privilege, has_any_column_privilege is false in redshift. 1) what!
+          "   pg_catalog.has_table_privilege(current_user, '\"' || t.schemaname || '\"' || '.' || '\"' || t.objectname || '\"',  'SELECT')"
+          "     OR pg_catalog.has_any_column_privilege(current_user, '\"' || t.schemaname || '\"' || '.' || '\"' || t.objectname || '\"',  'SELECT') as select,"
```
Comment (qnkhuat) [inline modules/drivers/redshift/src/metabase/driver/redshift.clj:446]: `has_any_column_privilege` return `false` if you have full select privileges :), we need to do `has_table_privilege OR has_any_column_privilege` here.

also has_any_column_privilege only exists on `select` and `update`.

File: modules/drivers/redshift/test/metabase/driver/redshift_test.clj [hunk 2]
```diff
@@ -400,76 +400,87 @@
...
+                        "REVOKE ALL PRIVILEGES ON SCHEMA \"%1$s\" FROM \"%7$s\";\n"
+                        "REVOKE USAGE ON SCHEMA \"%1$s\" FROM \"%7$s\";\n"
+                        "DROP USER IF EXISTS \"%7$s\";")
+                       schema-name
+                       qual-tbl-name
+                       qual-view-name
+                       qual-mview-name
+                       qual-tbl-partial-select-name
+                       qual-tbl-partial-update-name
+                       username)))))))))
```
Comment (calherries) [inline modules/drivers/redshift/test/metabase/driver/redshift_test.clj:486]: we'll need a newline here

File: modules/drivers/redshift/test/metabase/driver/redshift_test.clj [hunk 3]
```diff
@@ -400,76 +400,87 @@
...
-                       "GRANT SELECT ON %3$s TO \"%4$s\";")
+                       "CREATE TABLE %4$s (id INTEGER);\n"
+                       "CREATE TABLE %5$s (id INTEGER);\n"
+                       "CREATE USER \"%6$s\" WITH PASSWORD '%7$s';\n"
+                       "GRANT SELECT ON %1$s TO \"%6$s\";\n"
+                       "GRANT SELECT (id) ON %4$s TO \"%6$s\";\n"
+                       "GRANT UPDATE (id) ON %5$s TO \"%6$s\";\n"
+                       "GRANT UPDATE ON %1$s TO \"%6$s\";\n"
+                       "GRANT SELECT ON %2$s TO \"%6$s\";\n"
+                       "GRANT SELECT ON %3$s TO \"%6$s\";")
```
Comment (calherries) [inline modules/drivers/redshift/test/metabase/driver/redshift_test.clj:436]: having the grant statements in the same order as their corresponding args makes this easier to read imo
```suggestion
                       "GRANT UPDATE ON %1$s TO \"%6$s\";\n"
                       "GRANT SELECT ON %2$s TO \"%6$s\";\n"
                       "GRANT SELECT ON %3$s TO \"%6$s\";\n"
                       "GRANT SELECT (id) ON %4$s TO \"%6$s\";\n"
                       "GRANT UPDATE (id) ON %5$s TO \"%6$s\";")
```

**Review by calherries — APPROVED** [review]
Looks good, just left minor suggestions

File: modules/drivers/redshift/src/metabase/driver/redshift.clj (hunk 1 above)
Comment (calherries) [reply modules/drivers/redshift/src/metabase/driver/redshift.clj:446]: this is pretty surprising so let's leave this as a comment, like I've suggested here https://github.com/metabase/metabase/pull/40421/files#r1533731501

**Pull request merged**